### PR TITLE
Add tests for changes to Datadog.Trace's Public API

### PR DESCRIPTION
**Pull request by andrewlock**
I was thinking about our public API surface after the merge of the ClrProfiler assembly into Datadog.Trace, and thought it would be a good idea to start tracking changes to our public API in a deliberate way, especially with the further plans to extract some of this away in 2.0.

This PR uses [PublicApiGenerator](https://github.com/PublicApiGenerator/PublicApiGenerator) to assert that we don't (inadvertently) change our public API. 

As a test, I ran the diff for three points in time:

- `v1.28.2` tag (pre assembly merge)
- After the assembly merge ([full diff](https://github.com/DataDog/dd-trace-dotnet/commit/102a4825e22dcd9ee3b89d1029bb4a34e20464c8#diff-ecd517e4ded4d591bab4ac7c751e2cd3d8d321e37caee1462e824784187d3914))
- `master` latest commit ([full diff](https://github.com/DataDog/dd-trace-dotnet/commit/fad06e8d3c3e1e1798005480b7f7b343570c2b9f))

Some takeaways we should think about:
* All of our integrations were `public` in ClrProfiler.Managed which are _now_ public in Datadog.Trace, exposing a huge amount of "internal" auto-instrumentation infrastructure to users that are manually instrumenting their apps. I think we should be making those types all `internal` if possible
* The AppSec merged exposed some additional new types. We should confirm whether we want those to be public or not, and if not, make them `internal`.
* The API is different for each version we compile for. This is trivially because the `TargetFrameworkAttribute` is embedded, but also will be different if we have to publicly expose our auto-instrumentation classes. If we choose to make those internal, then I think we can (and should) assert that the APIs are identical across all platforms
* Verify doesn't support net45, so I excluded that from the tests

--- 

## Updates

I updated this PR to add `[Browsable(False)]` and `[EditorBrowsable(EditorBrowsableState.Never)]` to all the "public internal" types that were moved to Datadog.Trace.dll. This should hide them from intellisense for users that reference the nupkg (confirmed it definitely works in VS when you reference the assembly directly). The other "substantial" change I made was to make the `Security` class `internal`.

I also updated the public api tests to only test our _browsable_ API hasn't changed. This is something we definitely want to remain stable, so makes sense to be a unit test. 

The remaining question is: 
* Do we still want to document our _real_ public API (as opposed to our _effective_ API?
* How do we want to document it, given we expect it to change frequently? Could be a unit test, but that's a bit of a pain, given we _expect_ it to change often (e.g. every new integration). Could add a GitHub workflow to review and keep it up to date with each PR (had a go at this, but it's surprisingly annoying to do from inside Nuke)

The net result of the merge, after excluding all types marked with the attributes, is the much more palatable:

```diff
namespace Datadog.Trace.Configuration
         public const string AgentPort = "DD_TRACE_AGENT_PORT";
         public const string AgentUri = "DD_TRACE_AGENT_URL";
         public const string ApiKey = "DD_API_KEY";
+        public const string AppSecBlockingEnabled = "DD_APPSEC_BLOCKING_ENABLED";
+        public const string AppSecEnabled = "DD_APPSEC_ENABLED";
         public const string BufferSize = "DD_TRACE_BUFFER_SIZE";
         public const string ConfigurationFileName = "DD_TRACE_CONFIG_FILE";
         public const string CustomSamplingRules = "DD_TRACE_SAMPLING_RULES";
``` 

@DataDog/apm-dotnet

### CODE REVIEW

**Review by zacharycmontoya — APPROVED** [review]

**Review by tonyredondo — APPROVED** [review]
LGTM

File: src/Datadog.Trace/ClrProfiler/DynamicMethodBuilder.cs [hunk 1]
```diff
@@ -14,6 +15,8 @@ namespace Datadog.Trace.ClrProfiler
     /// </summary>
     /// <typeparam name="TDelegate">The type of delegate</typeparam>
     [Obsolete("This type will be removed in a future version of this library.")]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
     public static class DynamicMethodBuilder<TDelegate>
```
Comment (lucaspimentel) [inline src/Datadog.Trace/ClrProfiler/DynamicMethodBuilder.cs:20]: I think we can make this one `internal`.
```suggestion
    internal static class DynamicMethodBuilder<TDelegate>
```

File: src/Datadog.Trace/ClrProfiler/MemberAccessor.cs [hunk 2]
```diff
@@ -4,13 +4,16 @@
...
 
 namespace Datadog.Trace.ClrProfiler
 {
     /// <summary>
     /// Provides helper methods to access object members by emitting IL dynamically.
     /// </summary>
     [Obsolete("This type will be removed in a future version of this library.")]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
     public static class MemberAccessor
```
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace/ClrProfiler/MemberAccessor.cs:17]: Anyway, I think we can make it `internal`.
```suggestion
    internal static class MemberAccessor
```

**Review by lucaspimentel — CHANGES_REQUESTED** [review]
There are a few types than can probably be changed from `public` to `internal`. I'm not 100% so please test them before merging! 😅

**Comment by lucaspimentel** [comment]
Not sure what happened, but when I regenerated `integrations.json` on this branch, the entry for `AspNetCoreMiddlewareIntegration` was removed.

File: src/Datadog.Trace/ClrProfiler/IntegrationVersionRange.cs [hunk 3]
```diff
@@ -3,11 +3,15 @@
...
+using System.ComponentModel;
+
 namespace Datadog.Trace.ClrProfiler
 {
     /// <summary>
     /// Specifies a safe version range for an integration.
     /// </summary>
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
     public class IntegrationVersionRange
```
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace/ClrProfiler/IntegrationVersionRange.cs:15]: Good luck with StyleCop on this one 😆

File: src/Datadog.Trace/ClrProfiler/DynamicMethodBuilder.cs (hunk 1 above)
Comment (andrewlock) [reply src/Datadog.Trace/ClrProfiler/DynamicMethodBuilder.cs:20]: It's never used, and as it's now moved assembly, should we just remove it?

File: src/Datadog.Trace/ClrProfiler/MemberAccessor.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewlock) [reply src/Datadog.Trace/ClrProfiler/MemberAccessor.cs:17]: No, we're not using it anywhere. As it's moved assemblies, I think we can just remove it?

**Comment by andrewlock** [comment]
> I'm not 100% so please test them before merging!

Well, the tests all pass @lucaspimentel, so I guess we're ok! 😄

**Review by lucaspimentel — APPROVED** [review]
Thanks!

**Pull request merged**